rfcs/0155-worker-resource-monitoring.md

```diff
@@ -0,0 +1,135 @@
+existing data tools.
+
+## Motivation
+
+There is an active program to reduce costs in CI infrastructure, and more information is required about whether tasks are running on appropriately sized instances. The instance sizing also relates to developer efficiency and the time spent waiting for results, as well as the end-to-end time of releases.
+
+While AWS offers a [cost optimizer](https://console.aws.amazon.com/compute-optimizer/home?region=us-east-1#/dashboard), there are workers in GCP and Azure, requiring cloud-agnostic data collection.
+
+The worker should record samples of CPU times (system, user, iowait), Virtual and Swap memory usage, Available memory on the system, Disk IO and Network IO.
+
+For example:
+
+Task A is running on an workerType that provides 8 cores and 16Gb memory, and this will let us discover that Task A is single threaded, or only uses 8Gb memory. We can either then increase concurrency or reduce the instance size.
+
+Task B may be running on the same workerType as task A, but may use 1 core and 1.8Gb of memory, implying that a higher concurrency is effective.
+
+Task C may use the full memory of a worker, which may be efficient in terms of instance sizing but we could then highlight it as a relevant variable if the task needs to take less time.
+
+In all cases we would get a record of the resource impact of any payload changes, providing more information to developers.
+
+Since we already record task timing and cost data in BigQuery, that would be the ideal downstream location for this data. We already have an ingestion route for the live log in order to collect test suite timing data, and this could be the route for resource usage artifact ingestion.
+
+The downstream analysis would report on other tasks that would be affected by changes and inform whether changing an existing workerType's specification or creation of a new workerType is safer. We also plan on tooling to run experiments to measure the effects of changes.
+
+
+# Details
+
+Querying system-wide cpu and memory information isn't as useful, given that an instance can have multiple tasks running. So where possible, recording per-process stats is the goal. `psutil` and equivalents often have helper functions to allow quick discovery of all the process's children, and most platforms allow collection of per-process cpu times and memory usage.
+
+Disk and Network IO are valuable metrics to collect, with the caveats that only Linux has good support for these. Even querying per-process on Linux the data is system-wide, so we should be careful how that is presented, and not sum the data points along with the cpu times.
+
+
+
+Outline:
+
+1. Add code to the existing generic-worker and docker-worker
+2. Every 1 second, a sample will be taken:
+   a. Determine the number of processes being run by a task's payload
+   b. Sum CPU and memory usage for each of those processes
+   c. Record system-wide available memory at this timestamp
```

Comment:
Note that any system-wide data you get will contain data from other tasks if there are multiple tasks running on a given worker.

Reply:
Indeed, that's why I want to collect cpu and memory usage for the process tree, and only collect system-wide data if that's the only option.  Available memory is somewhat different as it lest us estimate memory pressure at the time

Reply:
OTOH, it might be useful to *also* have system-wide data for CPU and memory usage so as to know whether the pressure is because there are multiple tasks using too much resources at the same time.

Reply:
I've just double checked the timing information we're already collecting, and it doesn't store the instance ID, so we could either record just that and query for overlaps later on, or record extra samples at this stage. I'd lean towards the former, but could be convinced otherwise

Reply:
Would srfraser from 6 months from now remember to correlate with other tasks on the same instance? How about other people?

Reply:
I was imagining it would be part of the report, rather than a manual process each time. For our own use we'd want a pipeline for this that could be left alone and alert us when something looks like it could be acted on.

In any case, we're recording the total system memory, if we have available and "Memory I am using" then it's clear what's in use by other things, so it's a balance between collecting and storing extra data and how easy it is to work with. CPU times obviously don't work the same way, so it's a question of whether comparing system-wide cpu times to process-tree ones would show us anything actionable.